File: python/Heap or Priority Queue/kth_largest_element_stream.py
from typing import List
import heapq

class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        """
        Initialize the KthLargest class with a stream of numbers.
        Approach:
        - Maintain a min-heap of size k containing the k largest elements seen so far.
        - Heapify the initial list and pop the smallest until only k elements remain.
        Time Complexity: O(n log k) for init, O(log k) per add
        Space Complexity: O(k)
        """
        #MinHeap with K largest integers
        self.minHeap, self.k = nums, k
        heapq.heapify(self.minHeap)

        while len(self.minHeap) > k:
            heapq.heappop(self.minHeap)

    def add(self, val: int) -> int:
        """
        Add a new value to the stream and return the kth largest element.
        Time Complexity: O(log k)
        Space Complexity: O(k)
        """
        heapq.heappush(self.minHeap, val)
        
        if len(self.minHeap) > self. k:
            heapq.heappop(self.minHeap)
        return self.minHeap[0]

# -------------------
# Test Cases
# -------------------
def test_kth_largest_stream():
    # Test case 1: Basic
    kth = KthLargest(3, [4,5,8,2])
    assert kth.add(3) == 4  # [4,5,8], 2,3 dropped
    assert kth.add(5) == 5  # [5,5,8]
    assert kth.add(10) == 5 # [5,8,10]
    assert kth.add(9) == 8  # [8,9,10]
    assert kth.add(4) == 8  # [8,9,10]
    # Test case 2: k = 1
    kth2 = KthLargest(1, [])
    assert kth2.add(-3) == -3
    assert kth2.add(-2) == -2
    assert kth2.add(-4) == -2
    assert kth2.add(0) == 0
    assert kth2.add(4) == 4
    # Test case 3: All elements same
    kth3 = KthLargest(2, [2,2,2])
    assert kth3.add(2) == 2
    assert kth3.add(2) == 2
    # Test case 4: Negative numbers
    kth4 = KthLargest(2, [-1,-2,-3])
    assert kth4.add(-4) == -2
    assert kth4.add(0) == -1
    print("All test cases passed!")

if __name__ == "__main__":
    test_kth_largest_stream()

File: C++/Heap or Priority Queue/kth_largest_element_stream.cpp
#include <bits/stdc++.h>
using namespace std;

class KthLargest {
private:
    priority_queue<int, vector<int>, greater<int>> minHeap; // min-heap
    int k;

public:
    /**
     * Constructor
     * Initializes the KthLargest object with an integer k and a stream of integers nums.
     * Time: O(n log k) — inserting n elements into a heap of size at most k.
     * Space: O(k) — heap stores at most k elements.
     */
    KthLargest(int k, vector<int>& nums) {
        this->k = k;

        for (int num : nums) {
            minHeap.push(num);               // push all elements
            if (minHeap.size() > k) {
                minHeap.pop();              // maintain only k largest
            }
        }
    }

    /**
     * Adds an integer val to the stream and returns the k-th largest element.
     * Time: O(log k) — one push + one optional pop.
     * Space: O(k) — heap size stays at most k.
     */
    int add(int val) {
        minHeap.push(val);

        if (minHeap.size() > k) {
            minHeap.pop();                 // remove smallest if heap size > k
        }

        return minHeap.top();             // the k-th largest element
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers (e.g., vector, queue, etc.).

using namespace std;
    ✅ Allows usage of standard library names without prefixing std::.

class KthLargest { ... }
    ✅ Defines the KthLargest class.

private:
    ✅ Section for private members (accessible only within the class).

priority_queue<int, vector<int>, greater<int>> minHeap;
    ✅ Declares a min-heap (priority queue with smallest at top) using greater<int>.

int k;
    ✅ Stores the k value.

public:
    ✅ Section for public methods (accessible outside the class).

KthLargest(int k, vector<int>& nums)
    ✅ Constructor with two parameters: an int k and reference to a vector of ints.

this->k = k;
    ✅ Sets the class member k to the given k.

for (int num : nums)
    ✅ Range-based for loop iterates through nums vector.

minHeap.push(num);
    ✅ Inserts num into the minHeap.

if (minHeap.size() > k) minHeap.pop();
    ✅ If heap size exceeds k, remove smallest element.

int add(int val)
    ✅ Method that adds val to the stream and returns the k-th largest.

minHeap.push(val);
    ✅ Insert val into the heap.

if (minHeap.size() > k) minHeap.pop();
    ✅ Maintain heap size at most k.

return minHeap.top();
    ✅ Return smallest in heap, which is k-th largest overall.

*/
